util/src/main/java/org/datacommons/util/StatChecker.java

```diff
@@ -0,0 +1,351 @@
+    }
+    PercentDifference.Builder largestPercentDiff =
+        resBuilder.getSeriesLargestPercentDiffBuilder().setPercentDifference(maxDelta);
+    if (maxDeltaDP != null) {
+      largestPercentDiff.setDiffDataPoint(maxDeltaDP);
+    }
+    if (maxDeltaBaseDP != null) {
+      largestPercentDiff.setBaseDataPoint(maxDeltaBaseDP);
+    }
+    largestPercentDiff.build();
+  }
+
+  // Check if there are holes in the dates by inferring based on whether the successive dates have
+  // equal duration. If the sizes of the date strings don't match or there are invalid dates, don't
+  // try to infer holes and instead mark as "series_inconsistent_date_granularity" and
+  // "series_invalid_date".
+  protected static void checkDates(
+      List<DataPoint> timeSeries, StatValidationResult.Builder resBuilder) {
+    Set<LocalDateTime> dateTimes = new TreeSet<>();
+    StatValidationCounter.Builder invalidDateCounter = StatValidationCounter.newBuilder();
+    invalidDateCounter.setCounterName("Series_Invalid_Date");
+    // To keep track of the different lengths of the date strings.
+    Map<Integer, List<DataPoint>> dateLen = new HashMap<>();
+    // In the first pass, get sorted dates in LocalDateTime form and check for invalid dates and
+    // inconsistent date granularities.
+    for (DataPoint dp : timeSeries) {
+      String date = dp.getDate();
+      LocalDateTime dateTime = StringUtil.getValidISO8601Date(date);
+      if (dateTime == null) {
+        invalidDateCounter.addProblemPoints(dp);
+        continue;
+      }
+      if (!dateLen.containsKey(date.length())) {
+        dateLen.put(date.length(), new ArrayList<>());
+      }
+      dateLen.get(date.length()).add(dp);
+      dateTimes.add(dateTime);
+    }
+    List<Integer> dateLenList = new ArrayList<>(dateLen.keySet());
+    if (dateLenList.size() > 1) {
+      StatValidationCounter.Builder inconsistentDateCounter = StatValidationCounter.newBuilder();
+      inconsistentDateCounter.setCounterName("Series_Inconsistent_Date_Granularity");
+      // When there are multiple date granularity, the problem points will be those that are of a
+      // different date granularity than the most common one.
+      dateLenList.sort((d1, d2) -> dateLen.get(d2).size() - dateLen.get(d1).size());
+      for (int i = 1; i < dateLenList.size(); i++) {
+        dateLen.get(dateLenList.get(i)).forEach(inconsistentDateCounter::addProblemPoints);
+      }
+      resBuilder.addValidationCounters(inconsistentDateCounter.build());
+      return;
+    }
+    if (!invalidDateCounter.getProblemPointsList().isEmpty()) {
+      resBuilder.addValidationCounters(invalidDateCounter.build());
+      return;
+    }
+    long window = -1;
+    LocalDateTime prev = null;
+    List<LocalDateTime> dateTimesList = new ArrayList<>(dateTimes);
+    // In this second pass, compute the data holes.
+    for (LocalDateTime dt : dateTimesList) {
+      if (prev != null) {
+        long delta = ChronoUnit.MONTHS.between(prev, dt);
```

Comment:
I think prod also has a problem that we do not detect holes when granularity is weeks or days :( maybe add TODO?

Reply:
done